contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,826 @@
+as follows.
+    1. If the Pod's ordinal is in the sequence `[0,.Status.CurrentReplicas)`, 
+    the Pod should be consistent with version indicated by `Status.CurrentRevision`.
+    1. If the Pod's ordinal is in the sequence 
+    `[.Status.Replicas - .Status.UpdatedReplicas, .Status.Replicas)`
+    the Pod should be consistent with the version indicated by 
+    `Status.UpdateRevision`.
+1. If the StatefulSet's `.Spec.UpdateStrategy.Type` is equal to 
+`PartitionStatefulSetStrategyType` then the version of the Pod should be 
+as follows.
+    1. If the Pod's ordinal is in the sequence `[0,.Status.CurrentReplicas)`, 
+    the Pod should be consistent with version indicated by `Status.CurrentRevision`.
+    1. If the Pod's ordinal is in the sequence 
+    `[.Status.Replicas - .Status.UpdatedReplicas, .Status.Replicas)`the Pod 
+    should be consistent with the version indicated by `Status.UpdateRevision`.
+    1. If the Pod does not meet either of the prior two conditions, and if 
+    ordinal is in the sequence `[0, .Spec.UpdateStrategy.Partition.Ordinal)`, 
+    it should  be consistent with the version indicated by 
+    `Status.CurrentRevision`.
+    1. Otherwise, the Pod should be consistent with the version indicated 
+    by `Status.UpdateRevision`.
+
+### Pod State Reconciliation
+In order to reconcile a Pod with declared desired 
+[target state](#target-pod-state) the StatefulSet controller will do the 
+following.
+
+1. If the Pod is already consistent with its target state the controller will do 
+nothing.
+1. If the Pod is labeled with a `StatefulSetRevisionLabel` that indicates 
+the Pod was generated from a version of the StatefulSetSpec that is semantically 
+equivalent to, but not equal to, the [target version](#target-pod-state), the 
+StatefulSet controller will update the Pod with a `StatefulSetRevisionLabel` 
+indicating the new semantically equivalent version. This form of reconciliation 
+is non-destructive.
+1. If the Pod was not created from the target version, the Pod will be deleted 
+and recreated from that version. This form of reconciliation is destructive.
+
+### Specification Updates
+The StatefulSet controller will [snapshot](#snapshot-creation) its target 
+Object state when mutations are made to its `.Spec.Template` or 
+`.Spec.VolumeClaimsTemplate` (Note that the latter is currently immutable).
+
+1. When the StatefulSet controller observes a mutation to a StatefulSet's 
+ `.Spec.Template` it will snapshot its target Object state and compare 
+the snapshot with the version indicated by its `.Status.UpdateRevision`.
+1. If the current state is equivalent to the version indicated by 
+`.Status.UpdateRevision` no update has occurred. 
+1. If the `Status.CurrentRevision` field is empty, then the StatefulSet has no 
+revision history. To initialize its revision history, the StatefulSet controller 
+will set both `.Status.CurrentRevision` and `.Status.UpdateRevision` to the 
```

Comment:
Doesn't initializing `.Status.CurrentRevision` to the version of the current snapshot (which I understand to be a snapshot of the current target Object state) in this case introduce a conflict with the target pod state section?

That is, this rule says that `.Status.CurrentRevision == .Status.UpdateRevision`. Then `.Status.CurrentReplicas` starts off as 0 and `.Status.UpdatedReplicas` is also 0 (assuming target pod state is semantically different from the current pod state).

If upgrades proceed from pods with ordinals `.Replicas - 1 .. 0`, then `.Status.CurrentReplicas` and `.Status.UpdatedReplicas` will be incremented for each updated pod. But then pods in the sequence `[0, .Status.CurrentReplicas)` wouldn't be in the target pod state.